File: discipline_client_ts/discipline_android_primitive_ts/profile/application_name.ts
import { Nominal, TextualError, Tried } from "../x.ts";

const BRAND = Symbol();

export type ApplicationName = Nominal<typeof BRAND, string>;

export const MINIMUM_LENGTH = 1;
export const MAXIMUM_LENGTH = 30;

export const construct = (string: string): ApplicationName => {
  return Nominal.create(BRAND, string);
};

export const create = (string: string): Tried<ApplicationName, TextualError> => {
  if (string.length < MINIMUM_LENGTH) {
    const it = TextualError.create("Creating an ApplicationName from string");
    TextualError.addMessage(it, "String's length is less than the minimum valid length");
    TextualError.addStringAttachment(it, "String", string);
    TextualError.addNumberAttachment(it, "String length", string.length);
    TextualError.addNumberAttachment(it, "Minimum valid length", MINIMUM_LENGTH);
    return Tried.Failure(it);
  }

  if (string.length > MAXIMUM_LENGTH) {
    const it = TextualError.create("Creating an ApplicationName from string");
    TextualError.addMessage(it, "String's length is greater than the maximum allowed length");
    TextualError.addStringAttachment(it, "String", string);
    TextualError.addNumberAttachment(it, "String length", string.length);
    TextualError.addNumberAttachment(it, "Maximum valid length", MAXIMUM_LENGTH);
    return Tried.Failure(it);
  }

  return Tried.Success(construct(string));
};

export const ApplicationName = {
  create,
};

File: discipline4/app/src/main/java/com/example/app/rules/ApplicationName.kt
package com.yourpackage.discipline

import arrow.core.Either
import arrow.core.raise.either
import arrow.core.raise.raise

/**
 * Represents an application name with length constraints
 */
@JvmInline
value class ApplicationName private constructor(val value: String) {
    
    companion object {
        const val MINIMUM_LENGTH = 1
        const val MAXIMUM_LENGTH = 30
        
        /**
         * Creates an ApplicationName with validation
         */
        fun create(string: String): Either<TextualError, ApplicationName> = either {
            if (string.length < MINIMUM_LENGTH) {
                raise(
                    TextualError.create("Creating an ApplicationName from string")
                        .addMessage("String's length is less than the minimum valid length")
                        .addStringAttachment("String", string)
                        .addNumberAttachment("String length", string.length.toDouble())
                        .addNumberAttachment("Minimum valid length", MINIMUM_LENGTH.toDouble())
                )
            }
            
            if (string.length > MAXIMUM_LENGTH) {
                raise(
                    TextualError.create("Creating an ApplicationName from string")
                        .addMessage("String's length is greater than the maximum allowed length")
                        .addStringAttachment("String", string)
                        .addNumberAttachment("String length", string.length.toDouble())
                        .addNumberAttachment("Maximum valid length", MAXIMUM_LENGTH.toDouble())
                )
            }
            
            ApplicationName(string)
        }
        
        /**
         * Creates an ApplicationName without validation (use with caution)
         */
        fun construct(string: String): ApplicationName = ApplicationName(string)
    }
    
    override fun toString(): String = value
    
    fun isEqualTo(other: ApplicationName): Boolean = value == other.value
    
    fun length(): Int = value.length
}